doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,251 @@
+
+- To eliminate these duplicates, we will number the elements inside each group with 0, 1, 2, …, and this will be an additional component of the prefix sum indexes.
+- For this to perform well, we are assuming that groups are small.
+    - Note that a group here is identified by a value of the PARTITION BY expression + a value of the ORDER BY expression, so this is not an unreasonable assumption.
+- Also note that if there is no ORDER BY, then groups might be large, but in this case we don’t employ Prefix Sum, but we transform away the windowing to grouped aggregation + self-join (as noted above).
+
+**Parallelism.** DD's Prefix Sum should be data-parallel even inside a window partition. (It’s similar to [a Fenwick tree](https://en.wikipedia.org/wiki/Fenwick_tree), with sums maintained over power-of-2 sized intervals, from which you can compute a prefix sum by putting together LogN intervals.) TODO: But I wasn’t able to actually observe a speedup in a simple test when adding cores, so we should investigate what’s going on with parallelization. There was probably just some technical issue, because all operations in the Prefix Sum implementation look parallelizable.
+
+**Types.** We'll have to generalize DD's Prefix Sum to orderings over types other than a single unsigned integer, which is currently hardcoded in the code that forms the intervals. We’ll map other types to a single unsigned integer. Importantly, this mapping should *preserve the ordering* of the type:
+
+- Signed integer types are fine, we just need to fiddle with the sign to map them to an unsigned int in a way that preserves the ordering.
+- Date/Time types are just a few integers. We’ll concatenate their bits.
+- I don’t know how to handle strings, so these are out of scope for now. (Not seen in user queries yet.)
+
+## Alternatives
+
+There are several options for how to represent windowing in HIR, MIR, and LIR. For each of the IRs, I can see 3 options:
+
+1. Create a dedicated relation expression enum variant.
+2. Hide away windowing stuff in scalar expressions. (the current way in HIR)
+3. Reuse an existing relation expression enum variant, e.g., `Reduce`.
+
+I think we should do 1. for MIR and LIR, and later have a separate EPIC to consider unifying windowing and Reduce (and potentially TopK) into a many-to-many “super-reduce”.
+
+### Recognizing the current windowing pattern in just the rendering
+
+There is one more option that bypasses any MIR/LIR modifications: we could leave HIR, MIR, and LIR unchanged, and just make the rendering detect that pattern that the current HIR lowering creates, and do a custom rendering to it. However, this might be too brittle: we might accidentally leave out cases when the pattern somehow ends up slightly different, plus we might break it from time to time with unrelated changes. An example, for what gets created by the current HIR lowering from a window function call (`LAG`):
```

Comment:
Moving an open comment thread from Notion to here:

Frank:
I’m not sure I follow this argument. I’m pretty sure that we can (perhaps already do) have a scalar variant for window functions that we 100% have no clue how to transform, which ensures that we won’t accidentally remove it or transform it or anything.

Reply:
The current HIR lowering creates this entire big thing that the example here in the gray box shows:
- a Reduce with a 3-level nested row, 
- a FlatMap with an unnest_list,
- a Map with a bunch of record_gets,
- and some Projects around these.

This pattern does have a `lag(...)`, which is indeed a definitive indication that there is windowing going on. It’s also true that this lag definitely won’t get transformed away. However, to introduce a custom rendering based on Prefix Sum, it’s not enough to pattern-match just this lag, but we need to match the entire pattern of Reduce-Project-FlatMap-Map-Project.

One example complication in matching this pattern, is that MFP CSE can make the expressions of the Map look different from what the HIR lowering creates:
- If there is only one field being projected (besides the window function) at the end, then we will have a `record_get[xxx](record_get[1](#1))`
- But if we are projecting multiple fields, then MFP CSE will create a `record_get[1](#1)`, which goes into `#2`, and then there will be a `record_get[xxx](#2)` for each of the fields. It’s not impossible to match both of these cases, it’s just a bit cumbersome and possibly error-prone.

Reply:
In the meantime, we had a discussion in the office hours, and a new proposal was to create a many-to-many Reduce as a new MirRelationExpr enum variant. This would initially only contain windowing, but after the windowing is done, we would probably merge TopK into this.

Reply:
To loop back around, one imo totally valid proposal is to change nothing about the IR and use idiom detection in rendering, or potentially just in LIR. We already do this in several cases, most obviously with MFPs around operators that can be taught to absorb them (e.g. an MFP around a FlatMap, as in these examples).

The advantage this has is that you have already written correct implementations of projection pushdown, demand propagation, non-nullable requirements pushdown, the `typ()` function along with key detection and nullability of columns. All of this needs to be written anew for each variant that is a black box unrelated to the existing primitives. 

Looking at the code snippet below, the only defect seems to be that we don't absorb `Project` into `FlatMap`, which we should do in any case. At this point the idiom is "`FlatMap` with `unnest_list` atop `Reduce` with certain aggregates".

My understanding is that the specific reason this doesn't work for you is that someone needs to name a specific type for the ordering key, and the types are no longer visible in LIR if they are not captured. We could plausibly capture them in `lag` .. or each of the functions that supports the prefix sum representation.

My understanding is that nonetheless you would prefer to introduce new variants, but, as one of the people who needs to help fix all of the existing transforms, my default disposition is that you not introduce new variants, as it is more code to maintain.

Reply:
Answering this [here](https://github.com/MaterializeInc/materialize/pull/17077#issuecomment-1403712012).